java-frontend/src/main/java/org/sonar/java/ast/parser/JavaGrammar.java

```diff
@@ -157,12 +160,115 @@ public ModuleDeclarationTree MODULE_DECLARATION() {
+      .is(b.firstOf(
+        REQUIRES_MODULE_DIRECTIVE(),
+        EXPORTS_MODULE_DIRECTIVE(),
+        OPENS_MODULE_DIRECTIVE(),
+        USES_MODULE_DIRECTIVE(),
+        PROVIDES_MODULE_DIRECTIVE()));
+  }
+
+  public ModuleDirectiveTree REQUIRES_MODULE_DIRECTIVE() {
+    return b.<ModuleDirectiveTree>nonterminal(JavaLexer.REQUIRES_DIRECTIVE)
+      .is(b.firstOf(
+        // JLS9 - §3.9 : 'transitive' restricted keyword can be used as module name instead of modifier
+        f.newRequiresModuleDirective(
+          b.token(JavaRestrictedKeyword.REQUIRES),
+          b.token(JavaRestrictedKeyword.TRANSITIVE),
+          b.token(JavaPunctuator.SEMI)),
+        f.newRequiresModuleDirective(
+          b.token(JavaRestrictedKeyword.REQUIRES),
+          b.token(JavaKeyword.STATIC),
+          b.token(JavaRestrictedKeyword.TRANSITIVE),
+          b.token(JavaPunctuator.SEMI)),
+        // ordinary requires directives
+        f.newRequiresModuleDirective(
+          b.token(JavaRestrictedKeyword.REQUIRES),
+          b.zeroOrMore(REQUIRES_MODIFIER()),
+          MODULE_NAME(),
+          b.token(JavaPunctuator.SEMI))));
+  }
+
+  public InternalSyntaxToken REQUIRES_MODIFIER() {
+    return b.<InternalSyntaxToken>nonterminal(JavaLexer.REQUIRES_MODIFIER)
+      .is(b.firstOf(
+        b.token(JavaKeyword.STATIC),
+        b.token(JavaRestrictedKeyword.TRANSITIVE)));
+  }
+
+  public ModuleDirectiveTree EXPORTS_MODULE_DIRECTIVE() {
+    return b.<ModuleDirectiveTree>nonterminal(JavaLexer.EXPORTS_DIRECTIVE)
+      .is(
+        f.newExportsModuleDirective(
+          b.token(JavaRestrictedKeyword.EXPORTS),
+          EXPRESSION_QUALIFIED_IDENTIFIER(),
+          b.optional(
+            f.toModuleNames(
+              b.token(JavaRestrictedKeyword.TO),
+              MODULE_NAME_LIST())),
+          b.token(JavaPunctuator.SEMI)));
+  }
+
+  public ModuleDirectiveTree OPENS_MODULE_DIRECTIVE() {
+    return b.<ModuleDirectiveTree>nonterminal(JavaLexer.OPENS_DIRECTIVE)
+      .is(
+        f.newOpensModuleDirective(
+          b.token(JavaRestrictedKeyword.OPENS),
+          EXPRESSION_QUALIFIED_IDENTIFIER(),
+          b.optional(
+            f.toModuleNames2(
+              b.token(JavaRestrictedKeyword.TO),
+              MODULE_NAME_LIST())),
+          b.token(JavaPunctuator.SEMI)));
+  }
+
+  public ModuleDirectiveTree USES_MODULE_DIRECTIVE() {
+    return b.<ModuleDirectiveTree>nonterminal(JavaLexer.USES_DIRECTIVE)
+      .is(
+        f.newUsesModuleDirective(
+          b.token(JavaRestrictedKeyword.USES),
+          TYPE_QUALIFIED_IDENTIFIER(),
+          b.token(JavaPunctuator.SEMI)));
+  }
+
+  public ModuleDirectiveTree PROVIDES_MODULE_DIRECTIVE() {
+    return b.<ModuleDirectiveTree>nonterminal(JavaLexer.PROVIDES_DIRECTIVE)
+      .is(
+        f.newProvidesModuleDirective(
+          b.token(JavaRestrictedKeyword.PROVIDES),
+          EXPRESSION_QUALIFIED_IDENTIFIER(),
```

Comment:
Should be `TYPE_QUALIFIED_IDENTIFIER` to be consistent with `USES_MODULE_DIRECTIVE`